python/ray/serve/tests/test_cancellation.py

```diff
@@ -0,0 +1,418 @@
+    with pytest.raises(requests.exceptions.ReadTimeout):
+        requests.get("http://localhost:8000", timeout=0.5)
+
+    # Both the HTTP handler and the inner deployment handle call should be cancelled.
+    ray.get(inner_signal_actor.wait.remote())
+    ray.get(outer_signal_actor.wait.remote())
+
+
+@pytest.mark.skipif(
+    not RAY_SERVE_ENABLE_NEW_ROUTING,
+    reason="New routing feature flag is disabled.",
+)
+@pytest.mark.parametrize("use_fastapi", [False, True])
+def test_cancel_on_http_client_disconnect_during_assignment(
+    serve_instance, use_fastapi: bool
+):
+    """Test the client disconnecting while the proxy is assigning the request."""
+    signal_actor = SignalActor.remote()
+
+    @serve.deployment(max_concurrent_queries=1)
+    class Ingress:
+        def __init__(self):
+            self._num_requests = 0
+
+        async def __call__(self, *args):
+            self._num_requests += 1
+            await signal_actor.wait.remote()
+
+            return self._num_requests
+
+    h = serve.run(Ingress.bind()).options(use_new_handle_api=True)
+
+    # Send a request and wait for it to be ongoing so we know that further requests
+    # will block trying to assign a replica.
+    initial_response = h.remote()
+    wait_for_condition(lambda: ray.get(signal_actor.cur_num_waiters.remote()) == 1)
+
+    # Intentionally time out on the client, causing it to disconnect.
+    with pytest.raises(requests.exceptions.ReadTimeout):
+        requests.get("http://localhost:8000", timeout=0.5)
+
+    # Now signal the initial request to finish and check that the request sent via HTTP
+    # never reaches the replica.
+    ray.get(signal_actor.send.remote())
+    assert initial_response.result() == 1
+    for i in range(2, 12):
+        assert h.remote().result() == i
+
+
+@pytest.mark.skipif(
+    not RAY_SERVE_ENABLE_NEW_ROUTING,
+    reason="New routing feature flag is disabled.",
+)
+def test_cancel_sync_handle_call_during_execution(serve_instance):
+    """Test cancelling handle request during execution (sync context)."""
+    running_signal_actor = SignalActor.remote()
+    cancelled_signal_actor = SignalActor.remote()
+
+    @serve.deployment
+    class Ingress:
+        async def __call__(self, *args):
+            await running_signal_actor.send.remote()
+            await send_signal_on_cancellation(cancelled_signal_actor)
+
+    h = serve.run(Ingress.bind()).options(use_new_handle_api=True)
+
+    # Send a request and wait for it to start executing.
+    r = h.remote()
+    ray.get(running_signal_actor.wait.remote())
+
+    # Cancel it and verify that it is cancelled via signal.
+    r.cancel()
+    ray.get(cancelled_signal_actor.wait.remote())
+
+    with pytest.raises(ray.exceptions.TaskCancelledError):
+        r.result()
+
+
+@pytest.mark.skipif(
+    not RAY_SERVE_ENABLE_NEW_ROUTING,
+    reason="New routing feature flag is disabled.",
+)
+def test_cancel_sync_handle_call_during_assignment(serve_instance):
```

Comment:
Is the difference between this test and [`test_cancel_on_http_client_disconnect_during_assignment`](https://github.com/ray-project/ray/pull/38897/files#diff-60252887dd3f1ddc093ddbe90405b3632be95563e8add0a52ce75e25edff82cdR86) that this one uses Serve handles and the other one uses HTTP requests? That should be okay– I just want to make sure I'm not missing something.